🎉 New Source: 3PL Central

monai opened the pull request:
## What

- 3PL Central API source implementing the same streams as [tap-3plcentral](https://github.com/singer-io/tap-3plcentral).
- Participating in Hacktoberfest 2021, issue https://github.com/airbytehq/hacktoberfest-2021/issues/72


Upstream API documentation:

- https://api.3plcentral.com/Rels/
- https://developer.3plcentral.com/

## How

The source queries 3PL Central HTTP API using Python CDK and applies the following transformations:

- Converts all JSON object keys from `TitleCase` to `snake_case`.
- Recursively removes [HAL](https://stateless.group/hal_specification.html) `_link` keys from all objects.
- Adds primary key and cursor fields.

Unlike the Singer tap mentioned above, the source implementation does not apply any opinionated fetched data cleanup or flattening.

## Recommended reading order

1. `source.py`
2. `streams.py`
3.  `util.py`

## Pre-merge Checklist

#### Community member or Airbyter
   
- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [x] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

harshithmullapudi commented on airbyte-config/init/src/main/resources/seed/source_definitions.yaml:
```diff
@@ -351,6 +351,14 @@
+  dockerRepository: airbyte/source-tplcentral
+  dockerImageTag: dev
+  documentationUrl: https://docs.airbyte.io/integrations/sources/tplcentral
+- sourceDefinitionId: 4bfac00d-ce15-44ff-95b9-9e3c3e8fbd35
+  name: TikTok Marketing
+  dockerRepository: airbyte/source-tiktok-marketing
```
this is needed ?

monai replied:
Fixed.

harshithmullapudi commented on airbyte-integrations/connectors/source-tplcentral/acceptance-test-config.yml:
```diff
@@ -0,0 +1,30 @@
+    - config_path: "secrets/config.json"
+      configured_catalog_path: "integration_tests/configured_catalog.json"
+      empty_streams: []
+# TODO uncomment this block to specify that the tests should assert the connector outputs the records provided in the input file a file
+#      expect_records:
+#        path: "integration_tests/expected_records.txt"
+#        extra_fields: no
+#        exact_order: no
+#        extra_records: yes
```
this can't be done ?

monai replied:
While 3PL Central supports sandbox accounts, sadly, our account tier doesn't include it. In addition, I removed commented parts of the config.

harshithmullapudi commented on airbyte-integrations/connectors/source-tplcentral/integration_tests/sample_state.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "todo-stream-name": {
+    "todo-field-name": "value"
+  }
+}
```
remove this if this is not needed, if needed can you change this to actual stream name

monai replied:
Fixed.

harshithmullapudi commented on airbyte-integrations/connectors/source-tplcentral/source_tplcentral/source.py:
```diff
@@ -0,0 +1,118 @@
+        return [
+            StockSummaries(config),
+            Customers(config),
+
```
can you run `./gradlew format`

harshithmullapudi commented on airbyte-integrations/connectors/source-tplcentral/source_tplcentral/streams.py:
```diff
@@ -0,0 +1,268 @@
+
+
+class StockSummaries(TplcentralStream):
+    collection_field = "Summaries"
```
can you add comments about the URL for documentation to this stream

harshithmullapudi commented on airbyte-integrations/connectors/source-tplcentral/source_tplcentral/streams.py:
```diff
@@ -0,0 +1,268 @@
+            yield record
+
+
+class Customers(TplcentralStream):
```
can you add comments about the URL for documentation to this stream. Same for all the streams

harshithmullapudi reviewed (changes requested):
run `./gradlew format`

monai commented:
> run `./gradlew format`

@harshithmullapudi done

marcosmarxm commented:
Hello @monai! Sorry to not finished your contribution before the date stipulated in the contest. All contributions made before 15-November are eligible to receive the award. We're trying to review your contribution as soon as possible.

harshithmullapudi commented:
Hey @monai we are waiting for credentials here. We are yet to get those.

monai commented:
> Hey @monai we are waiting for credentials here. We are yet to get those.

Unfortunately, as I wrote to @marcosmarxm on Slack, we don't have a 3PL Central demo or limited account. However, I can share our production account credentials with you for a one-time test and rotate them immediately. That is, I can't provide account credentials for continuous integration.

alafanechere commented:
Hi @monai , 3LP is not very willing to create a sandbox account for us. Would you be able to provide us with creds on a test instance that we can use to make the test pass in our CI while we are negotiating with them?

monai commented:
Recent changes:

- Added schema source files to the repo since #8241 is fixed.
- Added two new properties to the schema that were added in 3PL Central API.
- Split configured catalog to one stream per file. Acceptance tests timed out due to long-running streams.
- Updated `documentationUrl` in `spec.json`.
- Run `gradlew format`.

harshithmullapudi reviewed (approved):
LGTM
1. Can you run `./gradlew format`
2. Also share the screenshots for integration_tests and unit_tests
3. Can you also share the cred on Slack (DM)

monai commented:
<details>
  <summary>Unit tests</summary>

```
~/projects/lt/airbyte/airbyte-integrations/connectors/source-tplcentral (source-tplcentral-alt) [um] python -m pytest -v unit_tests/                 [.venv]
Test session starts (platform: darwin, Python 3.9.7, pytest 6.1.2, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/juozas/projects/lt/airbyte, configfile: pytest.ini
plugins: cov-2.12.1, sugar-0.9.4, requests-mock-1.9.3, mock-3.6.1, timeout-1.4.2
collecting ...
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_cursor_field ✓                                  5% ▌
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_get_updated_state ✓                            10% ▉
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_stream_slices ✓                                14% █▌
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_supports_incremental ✓                         19% █▉
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_source_defined_cursor ✓                        24% ██▍
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py::test_stream_checkpoint_interval ✓                   29% ██▉
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_source.py::test_check_connection ✓                                          33% ███▍
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_source.py::test_streams ✓                                                   38% ███▊
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_request_params ✓                                           43% ████▍
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_next_page_token ✓                                          48% ████▊
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_next_page_token_with_page_size ✓                           52% █████▎
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_parse_response ✓                                           57% █████▊
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_parse_response_with_primary_key ✓                          62% ██████▎
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_parse_response_with_cursor_field ✓                         67% ██████▋
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_request_headers ✓                                          71% ███████▎
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_http_method ✓                                              76% ███████▋
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_should_retry[HTTPStatus.OK-False] ✓                        81% ████████▏
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_should_retry[HTTPStatus.BAD_REQUEST-False] ✓               86% ████████▋
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_should_retry[HTTPStatus.TOO_MANY_REQUESTS-True] ✓          90% █████████▏
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_should_retry[HTTPStatus.INTERNAL_SERVER_ERROR-True] ✓      95% █████████▌
 airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py::test_backoff_time ✓                                            100% ██████████
===================================================================== warnings summary ======================================================================
airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py: 6 warnings
airbyte-integrations/connectors/source-tplcentral/unit_tests/test_source.py: 6 warnings
airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py: 13 warnings
  /Users/juozas/projects/lt/airbyte/airbyte-integrations/connectors/source-tplcentral/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py:38: DeprecationWarning: Call to deprecated class NoAuth. (Set `authenticator=None` instead) -- Deprecated since version 0.1.20.
    self._authenticator = NoAuth()

airbyte-integrations/connectors/source-tplcentral/unit_tests/test_incremental_streams.py: 6 warnings
airbyte-integrations/connectors/source-tplcentral/unit_tests/test_source.py: 6 warnings
airbyte-integrations/connectors/source-tplcentral/unit_tests/test_streams.py: 13 warnings
  /Users/juozas/projects/lt/airbyte/airbyte-integrations/connectors/source-tplcentral/.venv/lib/python3.9/site-packages/deprecated/classic.py:173: DeprecationWarning: Call to deprecated class HttpAuthenticator. (Use requests.auth.AuthBase instead) -- Deprecated since version 0.1.20.
    return old_new1(cls, *args, **kwargs)

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (0.62s):
      21 passed
```

</details>

<details>
  <summary>Acceptance tests</summary>

```
~/projects/lt/airbyte/airbyte-integrations/connectors/source-tplcentral python -m pytest -p integration_tests.acceptance                                                                         [.venv]
Test session starts (platform: darwin, Python 3.9.7, pytest 6.1.2, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /Users/juozas/projects/lt/airbyte, configfile: pytest.ini
plugins: cov-2.12.1, sugar-0.9.4, requests-mock-1.9.3, mock-3.6.1, timeout-1.4.2
collecting ...
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                                                       4% ▍
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                                                             7% ▊
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                                                            11% █▏
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                                                          14% █▌
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓                                          18% █▊
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_defined_refs_exist_in_json_spec_file[inputs0] ✓                                21% ██▎
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                                               25% ██▌
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                                                         29% ██▉
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                                                         32% ███▎
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                                                       36% ███▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓                                39% ███▉
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_refs_exist_in_schema[inputs0] ✓                                   43% ████▍
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ✓                                                           46% ████▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs1] ✓                                                           50% █████
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs2] ✓                                                           54% █████▍
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs3] ✓                                                           57% █████▊
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs4] ✓                                                           61% ██████▏
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs5] ✓                                                           64% ██████▌
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                                     68% ██████▊
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs1] ✓                                     71% ███████▎
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                                  75% ███████▌
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs1] ✓                                  79% ███████▉
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs2] ✓                                  82% ████████▎
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs3] ✓                                  86% ████████▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] ✓                    89% ████████▉
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs1] ✓                    93% █████████▍
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs2] ✓                    96% █████████▋
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs3] ✓                   100% ██████████

Results (79.58s):
      28 passed
```

</details>
